plugincontainer/compatibility_matrix_test.go

```diff
@@ -40,28 +40,68 @@ func (m matrixInput) String() string {
 	return s
 }
 
+func withContainerRuntime(s string) func(matrixInput) matrixInput {
+	return func(m matrixInput) matrixInput {
+		m.containerRuntime = s
+		return m
+	}
+}
+
+func withRootlessEngine(b bool) func(matrixInput) matrixInput {
+	return func(m matrixInput) matrixInput {
+		m.rootlessEngine = b
+		return m
+	}
+}
+
+func withRootlessUser(b bool) func(matrixInput) matrixInput {
+	return func(m matrixInput) matrixInput {
+		m.rootlessUser = b
+		return m
+	}
+}
+
+func withMlock(b bool) func(matrixInput) matrixInput {
+	return func(m matrixInput) matrixInput {
+		m.mlock = b
+		return m
+	}
+}
+
 func TestCompatibilityMatrix(t *testing.T) {
 	if runtime.GOOS != "linux" {
 		t.Skip("Only linux is supported for now")
 	}
 
 	runCmd(t, "go", "build", "-o=examples/container/go-plugin-counter", "./examples/container/plugin-counter")
 
-	for _, runtime := range []string{runtimeRunc, runtimeRunsc} {
-		for _, rootlessEngine := range []bool{true, false} {
-			for _, rootlessUser := range []bool{true, false} {
-				for _, mlock := range []bool{true, false} {
-					i := matrixInput{
-						containerRuntime: runtime,
-						rootlessEngine:   rootlessEngine,
-						rootlessUser:     rootlessUser,
-						mlock:            mlock,
-					}
-					t.Run(i.String(), func(t *testing.T) {
-						runExamplePlugin(t, i)
-					})
-				}
-			}
+	testCases := [][]func(matrixInput) matrixInput{
+		{withRootlessEngine(true), withRootlessEngine(false)},
+		{withContainerRuntime(runtimeRunc), withContainerRuntime(runtimeRunsc)},
+		{withRootlessUser(true), withRootlessUser(false)},
+		{withMlock(true), withMlock(false)},
+	}
+	// Run a test for all combinations of 4 binary choices.
```

Comment:
I like this a lot. :)

I thought of a few slightly simpler ways of doing it, but they are minor optimizations at best, and probably not worth doing as they may or may not improve readability. (Like: switching to a Gray code to eliminate the `for`-`j` loop, or manipulating the boolean variables in `matrixInput` directly instead of using the functions.)

I also looked for some sort of Python `itertools.product`-like method that could do this for us, but sadly didn't find anything.

Reply:
Yeah a permutations/product method would make this pattern much easier to keep if we ever have more than 2 options for one of the inputs. I did make the functions a bit terser with inspiration from this comment, but otherwise agreed about other tweaks, even though they're fun :)